File: algorithm-interview/ch8/17/gina0605_swap_nodes_in_pairs.py
"""
24. Swap Nodes in Pairs
32ms, 14.3MB
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
	def swapPairs(self, head: ListNode) -> ListNode:
		p = ListNode(next=head)
		if head is not None and head.next is not None:
			head = head.next
		while p.next is not None and p.next.next is not None:
			q = p.next
			r = q.next
			s = r.next
			p.next = r
			r.next = q
			q.next = s
			p = q
		return head

File: algorithm-interview/ch8/17/gina0605_swap_nodes_in_pairs.cpp
/*
24. Swap Nodes in Pairs
4ms, 7.6MB
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
		ListNode *p = new ListNode(0, head);
		if(head != nullptr && head->next != nullptr) head = head->next;
		while(p->next != nullptr && p->next->next != nullptr) {
			ListNode *q = p->next;
			ListNode *r = q->next;
			ListNode *s = r->next;
			p->next = r;
			r->next = q;
			q->next = s;
			p = q;
		}
		return head;
    }
};
